File: public/mods/logicas/sala_jogos.js
// =====================================================
//   MECÂNICA: PONG MULTIPLAYER
//   Sala: sala_jogos
//   Servidor: server_mods/sala_jogos.py
// =====================================================
//
// API de plugin (ver mods/README_MODS.md):
//   onEnter(salaConfig)   — chamado ao entrar na sala
//   onSair()              — chamado ao sair da sala
//   onMensagem(dados, ws, meuBicho, tocarMusica, salaAtual)
//                         — retorne true se consumiu a mensagem
//   onTeclaDown(code, ws, meuBicho)
//                         — retorne true se consumiu a tecla
//   onFisica(meuBicho, ws, teclas)
//                         — retorne { bloqueiaMovimento, tremor }
//   render(ctx, meuBicho, outrosJogadores, imagensSprites, tamSprite)

SALA33_REGISTRAR("sala_jogos", {
    _jogando: false,
    _dados: { p1_y: 60, p2_y: 60, bola_x: 100, bola_y: 75, p1_score: 0, p2_score: 0 },
    _mesa: { x: 180, y: 195, w: 130, h: 30 },

    onEnter(salaConfig) {
        if (salaConfig?.extras?.mesaPong) this._mesa = salaConfig.extras.mesaPong;
        this._jogando = false;
    },

    onSair() {
        this._jogando = false;
    },

    onMensagem(dados, ws, meuBicho, tocarMusica, salaAtual) {
        if (dados.tipo !== "atualizacao_pong") return false;

        if (!this._jogando && dados.voce_esta_jogando) tocarMusica("pong");
        if (this._jogando && !dados.voce_esta_jogando) tocarMusica(salaAtual);

        this._jogando = dados.voce_esta_jogando;
        this._dados = dados.estado;

        if (dados.meu_x !== undefined) meuBicho.x = dados.meu_x;
        if (dados.meu_y !== undefined) meuBicho.y = dados.meu_y;
        if (dados.meu_lado !== undefined) meuBicho.lado = dados.meu_lado;
        return true;
    },

    onTeclaDown(code, ws, meuBicho) {
        if (code === "KeyE" && !this._jogando) {
            const m = this._mesa;
            const cx = meuBicho.x + meuBicho.tamanho / 2;
            const cy = meuBicho.y + meuBicho.tamanho / 2;
            if (cx > m.x - 20 && cx < m.x + m.w + 20 && cy > m.y - 20 && cy < m.y + m.h + 20) {
                ws.send(JSON.stringify({ tipo: "interagir_pong" }));
                return true;
            }
        }
        return false;
    },

    onFisica(meuBicho, ws, teclas) {
        if (!this._jogando) return { bloqueiaMovimento: false, tremor: 0 };
        if (teclas["ArrowUp"] || teclas["KeyW"]) ws.send(JSON.stringify({ tipo: "comando_pong", acao: "subir" }));
        if (teclas["ArrowDown"] || teclas["KeyS"]) ws.send(JSON.stringify({ tipo: "comando_pong", acao: "descer" }));
        if (teclas["KeyQ"]) ws.send(JSON.stringify({ tipo: "sair_pong" }));
        return { bloqueiaMovimento: true, tremor: 0 };
    },

    render(ctx, meuBicho, outrosJogadores, imagensSprites, tamSprite) {
        // Hint "[E] JOGAR PONG" quando próximo da mesa
        if (!this._jogando) {
            const m = this._mesa;
            const cx = meuBicho.x + tamSprite / 2;
            const cy = meuBicho.y + tamSprite / 2;
            if (cx > m.x - 20 && cx < m.x + m.w + 20 && cy > m.y - 20 && cy < m.y + m.h + 20) {
                ctx.fillStyle = "#161616"; ctx.fillRect(m.x + m.w/2 - 35, m.y - 22, 70, 14);
                ctx.strokeStyle = "#FFFFFF"; ctx.lineWidth = 1; ctx.strokeRect(m.x + m.w/2 - 35, m.y - 22, 70, 14);
                ctx.fillStyle = "#00ffcc"; ctx.font = "8px monospace"; ctx.textAlign = "center";
                ctx.fillText("[E] JOGAR PONG", m.x + m.w/2, m.y - 12);
            }
            return;
        }

        // Tela do Pong
        const d = this._dados;
        const px = 100, py = 40, pw = 200, ph = 140;
        ctx.fillStyle = "#000000"; ctx.fillRect(px, py, pw, ph);
        ctx.strokeStyle = "#FFFFFF"; ctx.lineWidth = 2; ctx.strokeRect(px, py, pw, ph);
        ctx.strokeStyle = "rgba(255,255,255,0.3)"; ctx.setLineDash([3, 3]);
        ctx.beginPath(); ctx.moveTo(px + pw/2, py); ctx.lineTo(px + pw/2, py + ph); ctx.stroke(); ctx.setLineDash([]);

        const escX = pw / 200, escY = ph / 150;
        ctx.fillStyle = "#FFFFFF";
        ctx.fillRect(px + 6 * escX, py + d.p1_y * escY, 4 * escX, 25 * escY);
        ctx.fillRect(px + (200 - 10) * escX, py + d.p2_y * escY, 4 * escX, 25 * escY);
        ctx.fillRect(px + d.bola_x * escX, py + d.bola_y * escY, 5 * escX, 5 * escY);
        ctx.font = "16px monospace"; ctx.textAlign = "center";
        ctx.fillText(d.p1_score, px + pw * 0.25, py + 25);
        ctx.fillText(d.p2_score, px + pw * 0.75, py + 25);
        ctx.font = "9px monospace"; ctx.fillStyle = "#ff2255";
        ctx.fillText("[Q] LARGAR CONTROLE", 200, py + ph + 15);
    },
});

File: server_mods/sala_jogos.py
# Mecânica: Pong Multiplayer (sala "sala_jogos")
# A física da bola roda no servidor a 60fps e o resultado é transmitido
# para todos os jogadores da sala.

import json
import random

HANDLES = ["interagir_pong", "comando_pong", "sair_pong"]
SALA = "sala_jogos"

STATE = {
    "p1_ws": None, "p2_ws": None,
    "p1_y": 60, "p2_y": 60,
    "bola_x": 100, "bola_y": 75,
    "bola_dx": 0, "bola_dy": 0,
    "p1_score": 0, "p2_score": 0,
    "ativo": False,
}


def _reset_bola():
    STATE["bola_x"] = 100
    STATE["bola_y"] = 75
    STATE["bola_dx"] = 2.5 if random.choice([True, False]) else -2.5
    STATE["bola_dy"] = 1.2 if random.choice([True, False]) else -1.2


def _reset_total():
    STATE["ativo"] = False
    STATE["p1_score"] = 0
    STATE["p2_score"] = 0
    STATE["p1_y"] = 60
    STATE["p2_y"] = 60
    STATE["bola_x"] = 100
    STATE["bola_y"] = 75
    STATE["bola_dx"] = 0
    STATE["bola_dy"] = 0


def on_leave(websocket, JOGADORES):
    if websocket == STATE["p1_ws"] or websocket == STATE["p2_ws"]:
        if websocket == STATE["p1_ws"]:
            STATE["p1_ws"] = None
        if websocket == STATE["p2_ws"]:
            STATE["p2_ws"] = None
        _reset_total()


async def tick(JOGADORES, SALAS, enviar_para_sala):
    if STATE["ativo"]:
        STATE["bola_x"] += STATE["bola_dx"]
        STATE["bola_y"] += STATE["bola_dy"]

        if STATE["bola_y"] <= 2 or STATE["bola_y"] >= 146:
            STATE["bola_dy"] *= -1

        if STATE["bola_x"] <= 10:
            if STATE["p1_y"] <= STATE["bola_y"] <= STATE["p1_y"] + 25:
                STATE["bola_dx"] = abs(STATE["bola_dx"]) * 1.05
                STATE["bola_dy"] = ((STATE["bola_y"] - (STATE["p1_y"] + 12.5)) / 12.5) * 2
            else:
                STATE["p2_score"] += 1
                _reset_bola()

        if STATE["bola_x"] >= 190:
            if STATE["p2_y"] <= STATE["bola_y"] <= STATE["p2_y"] + 25:
                STATE["bola_dx"] = -abs(STATE["bola_dx"]) * 1.05
                STATE["bola_dy"] = ((STATE["bola_y"] - (STATE["p2_y"] + 12.5)) / 12.5) * 2
            else:
                STATE["p1_score"] += 1
                _reset_bola()

    payload = {
        "tipo": "atualizacao_pong",
        "estado": {
            "p1_y": int(STATE["p1_y"]),
            "p2_y": int(STATE["p2_y"]),
            "bola_x": int(STATE["bola_x"]),
            "bola_y": int(STATE["bola_y"]),
            "p1_score": STATE["p1_score"],
            "p2_score": STATE["p2_score"],
        },
    }
    for ws in list(SALAS.get(SALA, set())):
        if ws in JOGADORES:
            payload["voce_esta_jogando"] = (ws == STATE["p1_ws"] or ws == STATE["p2_ws"])
            try:
                JOGADORES[ws]["queue"].put_nowait(json.dumps(payload))
            except Exception:
                pass


async def handle(tipo, websocket, dados, JOGADORES, SALAS, enviar_para_sala):
    if tipo == "interagir_pong":
        voce_sentou = False
        if not STATE["p1_ws"]:
            STATE["p1_ws"] = websocket
            JOGADORES[websocket]["x"] = 185
            JOGADORES[websocket]["y"] = 205
            JOGADORES[websocket]["lado"] = "direita"
            voce_sentou = True
        elif not STATE["p2_ws"] and STATE["p1_ws"] != websocket:
            STATE["p2_ws"] = websocket
            JOGADORES[websocket]["x"] = 270
            JOGADORES[websocket]["y"] = 205
            JOGADORES[websocket]["lado"] = "esquerda"
            voce_sentou = True

        if voce_sentou:
            await enviar_para_sala(SALA, {
                "tipo": "movimento", "id": id(websocket),
                "x": JOGADORES[websocket]["x"], "y": JOGADORES[websocket]["y"],
                "lado": JOGADORES[websocket]["lado"],
            })
            if STATE["p1_ws"] and STATE["p2_ws"] and not STATE["ativo"]:
                STATE["ativo"] = True
                STATE["p1_score"] = 0
                STATE["p2_score"] = 0
                _reset_bola()

            try:
                JOGADORES[websocket]["queue"].put_nowait(json.dumps({
                    "tipo": "atualizacao_pong",
                    "voce_esta_jogando": True,
                    "meu_x": JOGADORES[websocket]["x"],
                    "meu_y": JOGADORES[websocket]["y"],
                    "meu_lado": JOGADORES[websocket]["lado"],
                    "estado": {
                        "p1_y": int(STATE["p1_y"]), "p2_y": int(STATE["p2_y"]),
                        "bola_x": int(STATE["bola_x"]), "bola_y": int(STATE["bola_y"]),
                        "p1_score": STATE["p1_score"], "p2_score": STATE["p2_score"],
                    },
                }))
            except Exception:
                pass

    elif tipo == "comando_pong":
        acao = dados.get("acao")
        if websocket == STATE["p1_ws"]:
            if acao == "subir":
                STATE["p1_y"] = max(0, STATE["p1_y"] - 5)
            elif acao == "descer":
                STATE["p1_y"] = min(115, STATE["p1_y"] + 5)
        elif websocket == STATE["p2_ws"]:
            if acao == "subir":
                STATE["p2_y"] = max(0, STATE["p2_y"] - 5)
            elif acao == "descer":
                STATE["p2_y"] = min(115, STATE["p2_y"] + 5)

    elif tipo == "sair_pong":
        on_leave(websocket, JOGADORES)
        try:
            JOGADORES[websocket]["queue"].put_nowait(json.dumps({
                "tipo": "atualizacao_pong",
                "voce_esta_jogando": False,
                "estado": {"p1_y": 60, "p2_y": 60, "bola_x": 100, "bola_y": 75, "p1_score": 0, "p2_score": 0},
            }))
        except Exception:
            pass
